Switch to lookup table provider service

chris-allan opened the pull request:
# What this PR does

The previous lookup table loading implementation both leaves the task to
the individual service and has potentially severe performance
implications.  Two different loading mechanisms were used in the
thumbnail and rendering services due to the session state and service
availability.  Furthermore, the actual loading itself was deferred deep
within the renderer infrastructure.

This PR institutes a single `LutProvider` interface which can be
passed into services which desire access to the currently available
lookup tables.  A corresponding `LutProviderImpl` Spring Bean has also
been added, allowing for the implementation to be swapped in and out as
required.

Due to session containment differences between the thumbnail and
rendering services LUT loading was taken all the way.  Loading now
happens at server start time.  The only perceptible downside to this is
that adding a new LUT requires a server restart.  However, the
implementation was written in such a way that server runtime updates to
the available LUTs could be easily added if desired.

# Testing this PR

Use of the thumbnail and rendering services while both under and devoid of the influence of lookup tables should be unchanged.

# Related reading

Link to cards, tickets, other PRs:

1. https://trello.com/c/9iAe3Sg9/95-lut-support

2. glencoesoftware/omero-ms-image-region@8a95d980ea025168615cf5009d8349638b5c5e1d

chris-allan committed:
Switch to lookup table provider service

The previous lookup table loading implementation both leaves the task to
the individual service and has potentially severe performance
implications.  Two different loading mechanisms were used in the
thumbnail and rendering services due to the session state and service
availability.  Furthermore, the actual loading itself was deferred deep
within the renderer infrastructure.

This commit institutes a single `LutProvider` interface which can be
passed into services which desire access to the currently available
lookup tables.  A corresponding `LutProviderImpl` Spring Bean has also
been added, allowing for the implementation to be swapped in and out as
required.

Due to session containment differences between the thumbnail and
rendering services LUT loading was taken all the way.  Loading now
happens at server start time.  The only perceptible downside to this is
that adding a new LUT requires a server restart.  However, the
implementation was written in such a way that server runtime updates to
the available LUTs could be easily added if desired.

chris-allan committed:
Fix typo

chris-allan committed:
Unify style

chris-allan committed:
Fix and clean up unit tests

will-moore commented:
What's the method to use for adding LUTs to the server? Use existing API to add Original Files of ```mimetype = 'text/x-lut'```? In the user group? Or is the existing workflow of adding them to ```dist/lib/scripts/luts/``` still valid?

chris-allan committed:
Clean up javadoc warnings

chris-allan committed:
Change the LUT query strategy to utilize `ScriptRepoHelper`

During integration testing the sequence of Spring bean initialization
reared its ugly head.  In particular when a server is started for the
first time, none of the scripts have been registered in the database.
This largely invalidates the idea of using `SqlAction` directly and
forces us down the path of using `ScriptRepoHelper`.  This commit makes
that switch.

However, all is not sunshine and rainbows.  The scripts are actually
initialized, and by extension inserted into the database, by listening
for Spring Application Context refresh events.  We have to do the
same.  This has the side effect of "forcing" this initial script
initialization deeper down the stack into the "server" component and
effectively supplanting the work that `Ring.init()` does in checking for
script upates in the repository.

chris-allan commented:
@mtbc and @joshmoore will want to take a close look at 859cb6fd88eb73b7702bf231911b1432e481bddb and be sure we are happy with both the sequence and result of this change.

chris-allan commented:
There's no change at all the semantics of adding or removing LUTs, @will-moore. The only caveat is that we are not refreshing the list on every call to the `ThumbnailBean` or `RenderingBean`. A server restart is now required to get the current list of LUTs sync'd up with what is in the script repository.

chris-allan committed:
Remove superfluous continue

will-moore commented:
Several scripts test failures, due to this PR? https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-breaking-integration-java/855/

chris-allan committed:
Use direct filesystem strategy instead

There are just too many things, especially in the integration testing
infrastructure, that depend on the order of script insertion.  We can
safely use a direct filesystem strategy instead.  This improves
readability and reliability at the expense of slightly increasing the
possibility of a mismatch between the filesystem and the database.
However, this has always been, and will continue to be when utilising
the script infrastructure, a problem.

chris-allan commented:
One thing that had occurred to me is that #5273 does have the side effect of subtly breaking the API of `ScriptRepoHelper.loadAll(*)`. The return value of that method will now have the secret key inserted into the `name` attribute. We use the method so sparingly that I don't see this being a problem but it's probably something we should double check, @mtbc.

mtbc commented:
This morning's `LutProviderImpl.initLutReaders` exception makes me wonder if the problem was that the LUTScripts registration was occurring *before* `Ring.init` so the `_protect_originalfile_repo_*` database triggers had no valid `node.uuid` to check as a secret key.

chris-allan commented:
Certainly possible. There are so many potential races with all of this that I think the deferred initialization strategy is definitely the one to pursue. Should have something pushed with that change here shortly.

mtbc commented:
Might also suffice to add an explicit bean dependency in the Spring config but whichever works. :smiley:

chris-allan commented:
Yeah, I went down that road. Unfortunately, it quickly descends into bringing all these dependencies into the "blitz" component as a result. Which I did try but it felt really dirty. Especially as I would really like to be able to use these beans with the microservices and not have to initialize all of Blitz in order to do so. Furthermore, because the scripts are actually registered in the database by reacting to the Spring Application Context refresh events I'm not even 100% that it would work consistently. We already have enough problem with that as it is.

mtbc commented:
It's thus possible that relying on the Node UUIDs for allowing the updates to the file repo property was a design error that will need revisiting. In either case I concede that it's ugly.

chris-allan commented:
4e7ef8707fd6134a47839a83aa147467800c150e strips out all the weird stuff from 859cb6fd88eb73b7702bf231911b1432e481bddb and just goes directly to the filesystem in a `ScriptRepoHelper` compatible way. This gets rid of all of the concurrency issues, loads everything at startup time, and utilizes much of the same logic.

We have to go to the filesystem directly to load the LUTs anyway so I can't see a downside to this approach.

/cc @jburel

mtbc commented on components/server/src/ome/logic/LutProviderImpl.java:
```diff
@@ -0,0 +1,103 @@
+        for (ChannelBinding cb : channelBindings) {
+            if (cb.getActive()) {
+                LutReader lut = lutReaders.get(cb.getLookupTable());
+                if (lut != null) {
```
Seem to do the same thing in both cases?

chris-allan committed:
Simplify away name comparison hangover

joshmoore commented:
Comments from an initial read with Chris on Zoom:

 * one concern that may belong in this new `Provider` interface is the generation of the overview PNG
 * otherwise, looks like the default implementation could be replaced by one to use redis, which is nice
 * long-term, I'd like to see all the scripts cruft stripped out in favor of something more truly asynchronous

joshmoore commented:
@chris-allan points out that this could be a candidate for a milestone release due to the performance benefits, but that would require a general agreement on the behavior change related to reloading of LUTs.

cc: @sbesson @jburel

mtbc commented:
This PR causes https://trello.com/c/NSSdxI5n/64-bug-blocking-save-to-all-thumbs.

mtbc commented:
`org.hibernate.NonUniqueObjectException` exceptions attached to card.

joshmoore commented:
Excluding from develop to have 5.4.2 tested for a while without. Happy for this to be re-included on an integration branch.

jburel committed:
Fix savetoAll issue

mtbc commented:
That works excellently, thank you. Whatever the underlying issue is, the latest commit appears to work around it. :+1:

I wasted some time with the web client's "Save to All" saying "Cannot apply settings to other images". Turns out I had to move the images into a dataset -- pity the error gives no hint of that! Outwith this PR though.

jburel commented:
Discussed with @chris-allan, we will look at the underline issue post 5.4.4
Merging tomorrow unless any vetoes

chris-allan commented:
Sounds good to me!
